PROVIDER.md

```diff
@@ -1,168 +1,203 @@
-	return nil
-}
-```
-
-These methods are used by providers to read and store additional resources they may need during conversion.
-
-3. Create a struct named `converter` which implements the `ResourceConverter` interface in a file named `converter.go`.
-The implemented `ToGatewayAPI` function should simply call every registered `featureParser` function, one by one.
-Take a look at `ingressnginx/converter.go` for an example.
-The `ImplementationSpecificOptions` struct contains the handlers to customize native ingress implementation-specific fields.
-Take a look at `kong/converter.go` for an example.
-
-```go
-package examplegateway
-
-import (
-	"github.com/kubernetes-sigs/ingress2gateway/pkg/i2gw"
-)
-
-// converter implements the ToGatewayAPI function of i2gw.ResourceConverter interface.
-type converter struct {
-	conf *i2gw.ProviderConf
-
-	featureParsers []i2gw.FeatureParser
-	implementationSpecificOptions i2gw.ProviderImplementationSpecificOptions
-}
-
-// newConverter returns an ingress-nginx converter instance.
-func newConverter(conf *i2gw.ProviderConf) *converter {
-	return &converter{
-		conf: conf,
-		featureParsers: []i2gw.FeatureParser{
-			// The list of feature parsers comes here.
-		},
-		implementationSpecificOptions: i2gw.ProviderImplementationSpecificOptions{
-			// The list of the implementationSpecific ingress fields options comes here.
-		},
-	}
-}
-```
-4. Create a new struct named after the provider you are implementing. This struct should embed the previous 2 structs 
-you created.
-```go
-package examplegateway
-
-import (
-	"github.com/kubernetes-sigs/ingress2gateway/pkg/i2gw"
-)
-
-// Provider implements the i2gw.Provider interface.
-type Provider struct {
-	conf *i2gw.ProviderConf
-
-	*resourceReader
-	*converter
-}
-
-// NewProvider constructs and returns the example-gateway implementation of i2gw.Provider.
-func NewProvider(conf *i2gw.ProviderConf) i2gw.Provider {
-	return &Provider{
-		conf:           conf,
-		resourceReader: newResourceReader(conf),
-		converter:      newConverter(conf),
-	}
-}
-```
-5. Add the new provider to `i2gw.ProviderConstructorByName`.
-```go
-package examplegateway
-
-import (
-	"github.com/kubernetes-sigs/ingress2gateway/pkg/i2gw"
-)
-
-// The Name of the provider.
-const Name = "example-gateway-provider"
-
-func init() {
-	i2gw.ProviderConstructorByName[Name] = NewProvider
-}
-```
-6. Import the new package at `cmd/print`.
-```go
-package cmd
-
-import (
-	// Call init function for the providers
-	_ "github.com/kubernetes-sigs/ingress2gateway/pkg/i2gw/providers/ingressnginx"
-	_ "github.com/kubernetes-sigs/ingress2gateway/pkg/i2gw/providers/examplegateway"
-)
-```
+   `resource_reader.go`.
+
+   ```go
+   package examplegateway
+
+   import (
+       "context"
+
+       "github.com/kubernetes-sigs/ingress2gateway/pkg/i2gw/providers/common"
+   )
+
+   // converter implements the i2gw.CustomResourceReader interface.
+   type resourceReader struct {
+       conf *i2gw.ProviderConf
```

Comment:
@pottekkat I am not sure I understand what was the trigger for changing the docs ? 
There are some inaccurate bits like this, the package in the import has changed to `.../providers/common` (why?) but this `i2gw.ProviderConf` stayed.

I know you opened this PR a while ago (apologies it goes so long). Can you please go over the whole docs change and double check everything is aligned and make sense?

Reply:
@LiorLieberman The reason for changing the docs was the provider code was refactored. It might be outdated now and I have lost a lot of context. I have updated the PR to just add APISIX to the provider guide. I will get to updating the docs in a separate PR after going through the code again.